nexus/src/app/background/nat_cleanup.rs

```diff
@@ -0,0 +1,111 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background task for garbage collecting ipv4_nat_entry table.
+//! Responsible for cleaning up soft deleted entries once they
+//! have been propagated to running dpd instances.
+
+use super::common::BackgroundTask;
+use chrono::{Duration, Utc};
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::DataStore;
+use serde_json::json;
+use std::sync::Arc;
+
+/// Background task that periodically prunes soft-deleted entries
+/// from ipv4_nat_entry table
+pub struct Ipv4NatGarbageCollector {
+    datastore: Arc<DataStore>,
+    dpd_clients: Vec<Arc<dpd_client::Client>>,
+}
+
+impl Ipv4NatGarbageCollector {
+    pub fn new(
+        datastore: Arc<DataStore>,
+        dpd_clients: Vec<Arc<dpd_client::Client>>,
+    ) -> Ipv4NatGarbageCollector {
+        Ipv4NatGarbageCollector { datastore, dpd_clients }
+    }
+}
+
+impl BackgroundTask for Ipv4NatGarbageCollector {
+    fn activate<'a, 'b, 'c>(
+        &'a mut self,
+        opctx: &'b OpContext,
+    ) -> BoxFuture<'c, serde_json::Value>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        async {
+            let log = &opctx.log;
+
+            let result = self.datastore.ipv4_nat_current_version(opctx).await;
```

Comment:
To verify my understanding:

- This task periodically reads the latest generation number from the DB
- It sends that generation number to Dendrite instances (`client.ipv4_nat_generation()`)
- It cleans old records

So the "dendrite-playing-catch-up" entirely relies on Dendrite reaching back to the internal API within Nexus, correct?

Reply:
That's pretty close to how it works:
* Omicron periodically reads the latest generation in DB
* Omicron checks to see what generation numbers the various dendrite instances have (we are querying dendrite, not triggering any work in dendrite)
* Omicron only cleans up records older than the earliest found dendrite generation

The idea is that we don't clean up changes that a dendrite instance hasn't picked up yet. We also shouldn't have a detrimental TOCTOU race because the only way a dendrite instance moves backwards is if it crashes, and in that case it's okay that we're cleaning up old entries because we don't want that dendrite instance to re-apply them anyway.

Dendrite only catches up when:
* An event in the control plane calls the `trigger` endpoint on the dendrite API (a "doorbell" of sorts)
* Dendrite's internal RPW interval timer expires

In these events, yes, Dendrite will reach back via the internal API to catch up on the changes that have occurred.